### ISSUE TITLE
Tight loop in goroutine never yields 

### ISSUE THREAD

**Issue by 200sc** [gopherjs/gopherjs#698]
### What version of Go are you using (go version)?

go1.9 windows/amd64

### What version of GopherJS are you using ?

GopherJS 1.9-1

### What did you do?

I ran `gopherjs build` on [this program](https://github.com/oakmound/oak/tree/js/examples/pong) (with GOOS = darwin) and loaded it in my browser.

### What did you expect to see?

Well, I expected it to crash at some point or another as I've been struggling with getting this engine running in GopherJS for a while and incrementally fixing crashes as I find them. This crash happens to be out of my control without rewriting the content for building with js, or so it seems. 

I expected [this loop](https://github.com/oakmound/oak/blob/js/keyStore.go#L56) to give up its cpu time to other goroutines sometimes, as it does when built with go for windows or linux (I don't have a mac to test darwin, but as of maybe a month ago it also worked there).

### What did you see instead?

`Keyhold loop` never stops being printed, and all other goroutines cease advancing. (All fmt debug prints were added in order to find out where this hold up was, and so if `fmt.Println("Keyhold loop")` is removed, nothing changes.)

I understand the codebase for the engine is a bit much to parse through, so if requested I can attempt to put together a smaller demonstrative case.

**Comment by slimsag** [comment]
Curious: does adding a call to `runtime.Gosched()` inside that loop work around the issue?

**Comment by 200sc** [comment]
It might, I'll try it. We already do that in other tight loops that demonstrated this when built for desktop.

**Comment by flimzy** [comment]
>  It's possible that the difference in the scheduler just means that that's required in more places for gopherjs.

Yes. JavaScript runtimes are single-threaded, so the only time concurrent scheduling occurs is at function boundaries.  AFAIK, If your loop doesn't call any functions or do any blocking operations, it won't allow rescheduling.

**Comment by 200sc** [comment]
Well, wouldn't that depend on how gopherjs implements goroutines?

**Comment by flimzy** [comment]
> Well, wouldn't that depend on how gopherjs implements goroutines?

Yes, that is how GopherJS implements goroutines.  It breaks each function up into smaller chunks, at each call to a blocking operation. At these breakpoints, it allows other goroutines to execute.

But absent such breakpoints, as is the case in a tight loop with no function calls or blocking operations, that rescheduling never occurs.

**Comment by flimzy** [comment]
Looking at the [compiled output](https://gist.github.com/flimzy/15903655ad516cb4cb1f4a13530ea455) of that function, this can be confirmed; there are no `return`s found within the loop.

**Comment by flimzy** [comment]
[Here](https://gist.github.com/flimzy/15903655ad516cb4cb1f4a13530ea455) you can see the compiled version of the function in question.

**Comment by flimzy** [comment]
The fact that you're calling `Println` is interesting; that's a blocking function, so it should be enough to allow rescheduling.  Your "tight loop" doesn't actually fit the criteria I was describing above.

**Comment by neelance** [comment]
I can confirm that this is a limitation of GopherJS, there is no preemption. Switching to another goroutine only happens when code is blocking. There are no fairness guarantees. Adding this is impossible, as doing so would cripple performance too much. `runtime.Gosched()` should be a workaround.

**Comment by grantnelson-wf** [comment]
My 2 cents from 9 years down the road:

This matches what happens on the front-end when using await, async, and events in JS, Dart, TS, etc. Until a function hits a blocking call or a function boundary, the current events will continue running since there is only one thread. Some other applications similar to GopherJS, for example [goscript (Go→TS)](https://github.com/aperturerobotics/goscript) (still in progress) uses events in TS and has [the same issue](https://github.com/aperturerobotics/goscript/issues/118). Since the front-end is used to these kind of tight loop starvations, as far as I know, we still don't have a plan to fix this yet.

Some solutions to this problem are:

- compiling Go to JS LLVM instructions, but those can be slow. [TinyGo](https://tinygo.org/docs/guides/webassembly/) and [Go](https://go.dev/wiki/WebAssembly) have some WASM / WASI support but there are still some issues, limitations, and constraints with them at this time.

- something like mentioned above: `runtime.Gosched()`. Currently front-end languages will call something like `await sleep(0);` to let other pending events take a turn. This is similar to [µC/OS's context switch methods](https://micrium.atlassian.net/wiki/spaces/osiiidoc/pages/132394/OSCtxSw) that allows the developer to sprinkle manual interrupts in safe places (typically if a hardware interrupt is unavailable or disabled). If there is a need for something tighter, maybe we could add something like a `js.Sched()` that compiles to nothing, but tells GopherJS analyzer that there is a breaking statement at that location, thus it will compile in a scheduler check-in (a `$restore` switch case). However, although manually sprinkling those around the code gives the dev much better control, it may either not fix the starvation of a goroutine or could cause a lot of overhead.

- doing a deeper analysis on the [control-flow graph (CFG)](https://en.wikipedia.org/wiki/Control-flow_graph) model. The idea is to use the control-flow to find places to automatically inject context switches at a finer granularity than simply where required around blocking statements. However, from my experiments, this can cause slower resulting JS and is only an improvement as the resulting JS becomes closer to an LLVM's instructions. This would require a lot more research and work to get viable for GopherJS. My experiments are in a personal Go→TS project since haven't gotten far enough along to have anything consumable yet.

- use a webworker. You can compile into a separate JS that is loaded by a webworker and webworkers are asynchronous from the main JS. I did this in [the pending GopherJS playground update](https://github.com/gopherjs/gopherjs.github.io/pull/80) so that compiling is done in a webworker and it doesn't lock up the playground. The resulting JS is also run in a webworker to isolate it and allow a tight infinite loop to be killed via terminating the webworker. Obviously the issue is that webworkers don't have access to everything and you have to setup communication with it. However, if you have a large function for processing something, you can use it like a fast RPC where you send it arguments (and a unique id), then create a promise that is waiting for the error or result to return. The main draw back at the moment is that each compiled JS will carry some overlapping packages since each is self contained.
